### PR TITLE
Add tcp_backend

### LINKED ISSUE #58: Implement high-level API
This issue is solely meant for getting this off my mind.
We need to think of a high level API that handles connecting to other nodes, creating sockets and so on. 
For this we need to implementi a proper `middleman` and fitting `middleman_backend`s for specific protocols like `TCP`, `UDP` and so on.

### PR DESCRIPTION
**Pull request by jakobod**
This PR adds a `tcp_backend` and `publish` and `remote_actor` functions to the `middleman`.
closes #58 .

### CODE REVIEW

File: libcaf_net/src/net/backend/tcp.cpp [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    mm_.system().config(), "middleman.tcp-port", defaults::middleman::tcp_port);
+  ip_endpoint ep;
+  auto local_address = std::string("[::]:") + std::to_string(conf_port);
+  if (auto err = detail::parse(local_address, ep))
+    return err;
+  auto acceptor = make_tcp_accept_socket(ep, true);
+  if (!acceptor)
+    return acceptor.error();
+  auto acc_guard = make_socket_guard(*acceptor);
+  nonblocking(acc_guard.socket(), true);
```
Comment (Neverlord) [inline libcaf_net/src/net/backend/tcp.cpp:59]:
```suggestion
  if (auto err = nonblocking(acc_guard.socket(), true))
    return err;
```

File: libcaf_net/src/net/backend/tcp.cpp [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  return none;
+}
+
+void tcp::stop() {
+  for (const auto& p : peers_)
+    proxies_.erase(p.first);
+  peers_.clear();
+}
+
+expected<endpoint_manager_ptr> tcp::connect(const uri& locator) {
```
Comment (Neverlord) [inline libcaf_net/src/net/backend/tcp.cpp:85]: This function should first check whether the connection already exists.

File: libcaf_net/src/net/backend/tcp.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  return sec::cannot_connect_to_node;
+}
+
+endpoint_manager_ptr tcp::peer(const node_id& id) {
+  return get_peer(id);
+}
+
+void tcp::resolve(const uri& locator, const actor& listener) {
+  auto id = locator.authority_only();
+  if (id) {
```
Comment (Neverlord) [inline libcaf_net/src/net/backend/tcp.cpp:107]:
```suggestion
  if (auto id = locator.authority_only()) {
```

File: libcaf_net/src/net/backend/tcp.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  }
+  return sec::cannot_connect_to_node;
+}
+
+endpoint_manager_ptr tcp::peer(const node_id& id) {
+  return get_peer(id);
+}
+
+void tcp::resolve(const uri& locator, const actor& listener) {
+  auto id = locator.authority_only();
```
Comment (Neverlord) [inline libcaf_net/src/net/backend/tcp.cpp:106]:
```suggestion
```

File: libcaf_net/caf/net/backend/tcp.hpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+  void resolve(const uri& locator, const actor& listener) override;
+
+  strong_actor_ptr make_proxy(node_id nid, actor_id aid) override;
+
+  void set_last_hop(node_id*) override;
+
+  // -- properties -------------------------------------------------------------
+
+  tcp_stream_socket socket(const node_id& peer_id) {
```
[the remark this answers is not in the document]
Comment (jakobod) [reply libcaf_net/caf/net/backend/tcp.hpp:66]: We don't. I included it because the `test` backend has this function.

**Pull request merged**